### ISSUE TITLE
`unnecessary_final` tests should include variable patterns

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4121]
Child issue of https://github.com/dart-lang/linter/issues/3624 I suppose.

I think there should be a test case like:

```dart
switch ((1, 2)) {
  case (final a, final b): ...
}
```

in which the lint rule would report that var should be preferred.
